### PR TITLE
Add Email to Sign / In and Sign Up Page,  Settings Page, User can delete all Messages 

### PR DESCRIPTION
**Pull request by Gamius00**
- Added email field 
- user can create a delete all messages request and the other user can accept or reject 
- update userdata, user can update the own userdata (firstname, lastname, email)

https://github.com/FleetAdmiralJakob/chat-io/assets/127024349/bb83c724-8714-455f-a30e-a30905fe2928

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -56,7 +56,7 @@
     "sonner": "^1.5.0",
     "tailwind-merge": "^2.3.0",
     "tailwindcss-animate": "^1.0.7",
-    "zod": "^3.23.8"
+    "zod": "3.24.0-canary.20240523T174819"
```
Comment (FleetAdmiralJakob) [inline package.json:59]: Dont switch to canary without a reason

**Review by FleetAdmiralJakob — CHANGES_REQUESTED** [review]

File: src/app/(auth)/sign-up/signup-form.tsx [hunk 2]
```diff
@@ -244,6 +244,23 @@ export function SignUpForm() {
...
+          name="email"
+          render={({ field }) => (
+            <FormItem className="flex-2">
+              <FormLabel>Email</FormLabel>
+              <FormControl>
+                <Input placeholder="email" type="text" {...field} />
+              </FormControl>
+              <FormDescription>
+                This is optional but if you forgot your password we can send you
+                a Email
```
Comment (FleetAdmiralJakob) [inline src/app/(auth)/sign-up/signup-form.tsx:258]:
```suggestion
                a email
```
Typo

**Review by FleetAdmiralJakob — CHANGES_REQUESTED** [review]

File: src/app/(auth)/sign-up/signup-form.tsx [hunk 3]
```diff
@@ -244,6 +253,23 @@ export function SignUpForm() {
...
```
Comment (FleetAdmiralJakob) [inline src/app/(auth)/sign-up/signup-form.tsx:267]:
```suggestion
                an email.
```

Comment (FleetAdmiralJakob) [inline src/app/(auth)/sign-up/signup-form.tsx:267]:
```suggestion
                This is optional, but if you forgot your password, we can send you
                an email.
```

File: convex/chats.ts [hunk 4]
```diff
@@ -79,6 +79,7 @@ export const initialConvexSetup = mutation({
         username: identity.nickname,
         clerkId: identity.tokenIdentifier,
         firstName: identity.givenName,
+        email: identity.email,
```
Comment (Gamius00) [inline convex/chats.ts:82]: Set the value of the emailfield

File: convex/messages.ts [hunk 5]
```diff
@@ -39,6 +39,56 @@ export const getMessages = query({
   },
 });
 
+export const createDeleteRequest = mutation({
```
Comment (Gamius00) [inline convex/messages.ts:42]: This is the mutation for creating a delete all messages request

File: convex/messages.ts [hunk 6]
```diff
@@ -84,12 +134,62 @@ export const createMessage = mutation({
...
       privateChatId: parsedChatId,
       content: args.content.trim(),
+      type: "message",
       deleted: false,
       readBy: [convexUser._id],
     });
   },
 });
 
+export const deleteAllMessagesInChat = mutation({
```
Comment (Gamius00) [inline convex/messages.ts:144]: This is the mutation for delete all messages

File: convex/messages.ts [hunk 7]
```diff
@@ -84,12 +134,62 @@ export const createMessage = mutation({
...
+    const chat = ctx.table("privateChats").getX(parsedChatId);
+    const messagesInChat = await chat.edge("messages");
+
+    for (const message of messagesInChat) {
+      await message.delete();
+    }
+  },
+});
+
+export const rejectRequest = mutation({
```
Comment (Gamius00) [inline convex/messages.ts:169]: This is the reject request mutation

File: convex/schema.ts [hunk 8]
```diff
@@ -22,6 +23,7 @@ const schema = defineEntSchema({
 
   messages: defineEnt({})
     .field("content", v.string())
+    .field("type", v.string(), { default: "message" })
```
Comment (Gamius00) [inline convex/schema.ts:26]: This is the type of a message. 
I need a difference between the normal messages and the request messages

File: convex/users.ts [hunk 9]
```diff
@@ -13,3 +13,41 @@ export const getUserData = query({
     return ctx.table("users").getX("clerkId", identity.tokenIdentifier);
   },
 });
+
+export const updateUserData = mutation({
```
Comment (Gamius00) [inline convex/users.ts:17]: Here I update the user data (firstname, lastname, email)

File: src/app/(internal-sites)/chats/[chatId]/page.tsx [hunk 10]
```diff
@@ -104,6 +104,7 @@ export default function Page({ params }: { params: { chatId: string } }) {
         _creationTime: now,
         content,
         deleted: false,
+        type: "message",
```
Comment (Gamius00) [inline src/app/(internal-sites)/chats/[chatId]/page.tsx:107]: This is the normal type of any message that is not a request

File: src/app/(internal-sites)/profile/page.tsx [hunk 11]
```diff
@@ -1,73 +1,114 @@
...
-          <div className="ml-4 text-sm xl:mt-2 xl:text-lg">
-            {user?.lastName && user.firstName ? (
-              <div>
-                {user.firstName} {user?.lastName} / <br className="xl:hidden" />{" "}
-                {user.username}
-              </div>
-            ) : (
-              <div className="xl:mt-4">{user?.username}</div>
-            )}
+        <p className="my-10 text-xl font-bold sm:text-2xl">Profile</p>
```
Comment (Gamius00) [inline src/app/(internal-sites)/profile/page.tsx:33]: This is only the profilepage

File: src/app/api/sign-up/route.ts [hunk 12]
```diff
@@ -19,18 +38,35 @@ export async function POST(request: Request) {
...
     if (isClerkAPIResponseError(e)) {
       if (e.errors.some((error) => error.code === "form_identifier_exists")) {
-        return Response.json(
-          {
-            message: "Failed to create an account. Username already exists.",
-            statusText: "username_is_taken",
-          },
-          { status: 400 },
-        );
+        if (
```
Comment (Gamius00) [inline src/app/api/sign-up/route.ts:49]: Added a email error

File: src/components/message.tsx [hunk 13]
```diff
@@ -274,11 +327,47 @@ export const Message = ({
                   <Ban />
                   <p className="ml-2.5">This message was deleted</p>
                 </div>
+              ) : message.type != "message" ? (
```
Comment (Gamius00) [inline src/components/message.tsx:330]: Added the UI of the delete messages request

File: tailwind.config.ts [hunk 14]
```diff
@@ -25,6 +25,7 @@ const config = {
           dark: "#005C8F",
           white: "#009BFA",
         },
+        accept: "#00834C",
```
Comment (Gamius00) [inline tailwind.config.ts:28]: Added new color for the accept button

File: package.json [hunk 15]
```diff
@@ -56,7 +56,7 @@
     "sonner": "^1.5.0",
     "tailwind-merge": "^2.3.0",
     "tailwindcss-animate": "^1.0.7",
-    "zod": "^3.23.8"
+    "zod": "3.23.8"
```
Comment (FleetAdmiralJakob) [inline package.json:59]: Make it as it was before (^^^^^^)

**Review by FleetAdmiralJakob — CHANGES_REQUESTED** [review]

File: src/components/password-dialog.tsx
Comment (FleetAdmiralJakob) [inline src/components/password-dialog.tsx:1]: Whats this file doing? Why not use the same dialog as for the add user dialog?

**Review by FleetAdmiralJakob — APPROVED** [review]

**Review by FleetAdmiralJakob — APPROVED** [review]

**Review by FleetAdmiralJakob — APPROVED** [review]

**Pull request merged**